File: Kotlin/ch01/src/main/kotlin/ch01/LinkedList.kt
package ch01

/**
 * 단순 연결 리스트 (Singly Linked List)
 * Rust/ch01/src/linked_list.rs → Kotlin 이디엄 포팅
 */
class LinkedList {

    class Node(val data: Int) {
        var next: Node? = null
    }

    private var head: Node? = null

    /** 리스트 끝에 노드를 추가한다. */
    fun append(data: Int): Node {
        val node = Node(data)
        if (head == null) {
            head = node
        } else {
            var tail = head!!
            while (tail.next != null) tail = tail.next!!
            tail.next = node
        }
        return node
    }

    /** 현재 노드 뒤에 새 노드를 삽입한다. */
    fun insertAfter(current: Node, data: Int): Node {
        val node = Node(data)
        node.next = current.next
        current.next = node
        return node
    }

    /** 리스트 앞에 새 헤드 노드를 삽입한다. */
    fun insertHead(data: Int): Node {
        val node = Node(data)
        node.next = head
        head = node
        return node
    }

    /** 노드를 리스트에서 분리한다 (메모리는 GC가 해제). */
    fun remove(node: Node) {
        if (head === node) {
            head = node.next
        } else {
            var current = head
            while (current?.next !== node) current = current?.next
            current?.next = node.next
        }
        node.next = null
    }

    /** index 위치의 노드를 반환한다. 범위를 벗어나면 null. */
    fun getAt(index: Int): Node? {
        if (index < 0) return null
        var current = head
        repeat(index) { current = current?.next }
        return current
    }

    /** 노드 개수 */
    val count: Int
        get() {
            var n = 0
            var cur = head
            while (cur != null) { n++; cur = cur.next }
            return n
        }

    /** 리스트를 Int 리스트로 변환 */
    fun toList(): List<Int> = buildList {
        var cur = head
        while (cur != null) { add(cur.data); cur = cur.next }
    }
}

fun main() {
    val list = LinkedList()

    // 노드 5개 추가
    for (i in 0 until 5) list.append(i)
    list.insertHead(-1)
    list.insertHead(-2)

    // 리스트 출력
    for (i in 0 until list.count) println("List[$i] : ${list.getAt(i)!!.data}")

    // 리스트의 세 번째 노드 뒤에 새 노드 삽입
    println("\nInserting 3000 After [2]...\n")
    list.insertAfter(list.getAt(2)!!, 3000)

    for (i in 0 until list.count) println("List[$i] : ${list.getAt(i)!!.data}")

    // 모든 노드를 리스트에서 제거
    println("\nDestroying List...")
    while (list.count > 0) list.remove(list.getAt(0)!!)
}

File: Rust/ch01/src/linked_list.rs
// 단순 연결 리스트 (Singly Linked List)
// Clang/01/LinkedList 포팅. C의 포인터 기반 구조를 raw pointer로 충실히 옮겼다.
use std::ptr;

pub type ElementType = i32;

pub struct Node {
    pub data: ElementType,
    pub next: *mut Node,
}

pub unsafe fn sll_create_node(new_data: ElementType) -> *mut Node {
    Box::into_raw(Box::new(Node {
        data: new_data,
        next: ptr::null_mut(),
    }))
}

pub unsafe fn sll_destroy_node(node: *mut Node) {
    drop(Box::from_raw(node));
}

pub unsafe fn sll_append_node(head: &mut *mut Node, new_node: *mut Node) {
    if head.is_null() {
        *head = new_node;
    } else {
        let mut tail = *head;
        while !(*tail).next.is_null() {
            tail = (*tail).next;
        }
        (*tail).next = new_node;
    }
}

pub unsafe fn sll_insert_after(current: *mut Node, new_node: *mut Node) {
    (*new_node).next = (*current).next;
    (*current).next = new_node;
}

pub unsafe fn sll_insert_new_head(head: &mut *mut Node, new_head: *mut Node) {
    if head.is_null() {
        *head = new_head;
    } else {
        (*new_head).next = *head;
        *head = new_head;
    }
}

pub unsafe fn sll_remove_node(head: &mut *mut Node, remove: *mut Node) {
    if *head == remove {
        *head = (*remove).next;
    } else {
        let mut current = *head;
        while !current.is_null() && (*current).next != remove {
            current = (*current).next;
        }
        if !current.is_null() {
            (*current).next = (*remove).next;
        }
    }
}

pub unsafe fn sll_get_node_at(head: *mut Node, mut location: i32) -> *mut Node {
    let mut current = head;
    if location < 0 {
        return ptr::null_mut();
    }
    while !current.is_null() && location > 0 {
        current = (*current).next;
        location -= 1;
    }
    current
}

pub unsafe fn sll_get_node_count(head: *mut Node) -> i32 {
    let mut count = 0;
    let mut current = head;
    while !current.is_null() {
        current = (*current).next;
        count += 1;
    }
    count
}

#[cfg(test)]
mod tests {
    use super::*;

    // 테스트 헬퍼: 리스트를 Vec로 수집한다.
    unsafe fn to_vec(head: *mut Node) -> Vec<ElementType> {
        let mut out = Vec::new();
        let mut current = head;
        while !current.is_null() {
            out.push((*current).data);
            current = (*current).next;
        }
        out
    }

    unsafe fn destroy(head: &mut *mut Node) {
        while !head.is_null() {
            let node = *head;
            *head = (*node).next;
            sll_destroy_node(node);
        }
    }

    #[test]
    fn append_builds_in_order() {
        unsafe {
            let mut list: *mut Node = ptr::null_mut();
            for i in 0..5 {
                sll_append_node(&mut list, sll_create_node(i));
            }
            assert_eq!(sll_get_node_count(list), 5);
            assert_eq!(to_vec(list), vec![0, 1, 2, 3, 4]);
            destroy(&mut list);
        }
    }

    #[test]
    fn insert_new_head_prepends() {
        unsafe {
            let mut list: *mut Node = ptr::null_mut();
            for i in 0..3 {
                sll_append_node(&mut list, sll_create_node(i));
            }
            sll_insert_new_head(&mut list, sll_create_node(-1));
            assert_eq!(to_vec(list), vec![-1, 0, 1, 2]);
            destroy(&mut list);
        }
    }

    #[test]
    fn insert_after_middle() {
        unsafe {
            let mut list: *mut Node = ptr::null_mut();
            for i in 0..3 {
                sll_append_node(&mut list, sll_create_node(i));
            }
            let second = sll_get_node_at(list, 1);
            sll_insert_after(second, sll_create_node(99));
            assert_eq!(to_vec(list), vec![0, 1, 99, 2]);
            destroy(&mut list);
        }
    }

    #[test]
    fn remove_head_and_middle() {
        unsafe {
            let mut list: *mut Node = ptr::null_mut();
            for i in 0..4 {
                sll_append_node(&mut list, sll_create_node(i));
            }
            // 헤드 제거
            let head = sll_get_node_at(list, 0);
            sll_remove_node(&mut list, head);
            sll_destroy_node(head);
            assert_eq!(to_vec(list), vec![1, 2, 3]);
            // 가운데 제거
            let mid = sll_get_node_at(list, 1);
            sll_remove_node(&mut list, mid);
            sll_destroy_node(mid);
            assert_eq!(to_vec(list), vec![1, 3]);
            destroy(&mut list);
        }
    }

    #[test]
    fn get_node_at_out_of_range() {
        unsafe {
            let mut list: *mut Node = ptr::null_mut();
            sll_append_node(&mut list, sll_create_node(7));
            assert!(sll_get_node_at(list, -1).is_null());
            assert!(sll_get_node_at(list, 5).is_null());
            destroy(&mut list);
        }
    }
}
